### DISCUSSION TITLE
Reading PDF file

### DISCUSSION THREAD

**Discussion by e1imba** [dlemstra/Magick.NET#990] [Help]
Hi @dlemstra, first I want to thank you for this great library. 

I am trying to read data from PDF and similar file types. 
I saw that I can get xmp profile with GetXmpProfile() method but I am wondering is there any way to extract comments from file with your library?

**Comment by dlemstra** [comment]
This library is a not PDF parser. If you want to read specific PDF properties you might want to look at another library. It is also possible that we read the information that you need. Is it possible to share a file and the comment that you are looking for from that file?

**Reply by e1imba** [reply]
Thank you for answer. I saw that I can read XMP profile, so I though, maybe I can do same with comments :) 
My last goal is to read custom colors from PDF file, and I saw that I have that information in XMP Profile and in comments.

I will share only certain parts of PDF.

%%DocumentFonts: 
%%DocumentNeededFonts: 
%%DocumentNeededFeatures: 
%%DocumentSuppliedFeatures: 
%%DocumentProcessColors:  Black
**%%DocumentCustomColors: (PANTONE 320 U)
%%CMYKCustomColor: 0.9326 0.1530 0.3581 0.0010 (PANTONE 320 U)**
%%RGBCustomColor: 
%%EndComments

I see that I have this information in XMP profile, but I was not sure is that only option to get custom colors from Magick.NET library. For now, my option is to parse XMP profile to get this data. If you have better idea, I will accept it with pleasure.

<xmpTPg:PlateNames>
            <rdf:Seq>
               <rdf:li>Black</rdf:li>
               <rdf:li>PANTONE 320 U</rdf:li>
            </rdf:Seq>
         </xmpTPg:PlateNames>
         <xmpTPg:SwatchGroups>
            <rdf:Seq>
               <rdf:li rdf:parseType="Resource">
                  <xmpG:groupName>Default Swatch Group</xmpG:groupName>
                  <xmpG:groupType>0</xmpG:groupType>
                  <xmpG:Colorants>
                     <rdf:Seq>
                        <rdf:li rdf:parseType="Resource">
                           <xmpG:swatchName>White</xmpG:swatchName>
                           <xmpG:mode>CMYK</xmpG:mode>
                           <xmpG:type>PROCESS</xmpG:type>
                           <xmpG:cyan>0.000000</xmpG:cyan>
                           <xmpG:magenta>0.000000</xmpG:magenta>
                           <xmpG:yellow>0.000000</xmpG:yellow>
                           <xmpG:black>0.000000</xmpG:black>
                        </rdf:li>
                        <rdf:li rdf:parseType="Resource">
                           <xmpG:swatchName>Black</xmpG:swatchName>
                           <xmpG:mode>CMYK</xmpG:mode>
                           <xmpG:type>PROCESS</xmpG:type>
                           <xmpG:cyan>0.000000</xmpG:cyan>
                           <xmpG:magenta>0.000000</xmpG:magenta>
                           <xmpG:yellow>0.000000</xmpG:yellow>
                           <xmpG:black>100.000000</xmpG:black>
                        </rdf:li>
                        <rdf:li rdf:parseType="Resource">
                           <xmpG:swatchName>PANTONE 320 U</xmpG:swatchName>
                           <xmpG:type>SPOT</xmpG:type>
                           <xmpG:tint>100.000000</xmpG:tint>
                           <xmpG:mode>CMYK</xmpG:mode>
                           <xmpG:cyan>93.258560</xmpG:cyan>
                           <xmpG:magenta>15.298694</xmpG:magenta>
                           <xmpG:yellow>35.809872</xmpG:yellow>
                           <xmpG:black>0.091554</xmpG:black>
                        </rdf:li>
                        <rdf:li rdf:parseType="Resource">
                           <xmpG:swatchName>PANTONE 323 U</xmpG:swatchName>
                           <xmpG:type>SPOT</xmpG:type>
                           <xmpG:tint>100.000000</xmpG:tint>
                           <xmpG:mode>LAB</xmpG:mode>
                           <xmpG:L>37.647060</xmpG:L>
                           <xmpG:A>-16</xmpG:A>
                           <xmpG:B>-9</xmpG:B>
                        </rdf:li>
                     </rdf:Seq>
                  </xmpG:Colorants>

Your library suits perfect with all other needs, and I am looking to not use other libraries only for getting custom colors.

**Reply by dlemstra** [reply]
It looks like we are reading those two values as attributes on the image. You might be able to get the information with: `image.GetAttribute($"pdf:SpotColor-{i}`. Where `i` is a value starting at `0` and as high as the number of spot color information (`-1`) we find. I am unable to test this since I don't have a PDF document with that information.

Another option would be to parse the document yourself. That will probably be faster if you just need that information.

**Reply by e1imba** [reply]
I tried with pdf:SpotColor-{i}, but I am getting only name of parameters/attributes (DocumentCustomColors, CMYKCustomColor) not exactly data ( PANTONE 320 U , 0.9326 0.1530 0.3581 0.0010 ).

I am already working on parsing the document, but if I can get it from GetAttribute() method it would be very nice.

**Reply by dlemstra** [reply]
It looks like there is a bug in the current code of ImageMagick that tries to read those values. With a fix for that bug I get the following output with the PDF file that you emailed me (thank you!):

```
pdf:SpotColor-0: rocset Adobe_CoolType_Utility_T42 1.0 0
pdf:SpotColor-1: rocset Adobe_CoolType_Utility_MAKEOCF 1.23 0
pdf:SpotColor-2: rocset Adobe_CoolType_Core 2.31 0
pdf:SpotColor-3: rocset Adobe_AGM_Core 2.0 0
pdf:SpotColor-4: rocset Adobe_AGM_Utils 1.0 0
pdf:SpotColor-5: PANTONE 320 U
pdf:SpotColor-6: 0.9326 0.1530 0.3581 0.0010 PANTONE 320 U
```

This will be resolved in the next version of ImageMagick/Magick.NET.

**Reply by e1imba** [reply]
Cool. Thank you!
